Bluetooth: Mesh: Add artificial beacon delay

Andrewpini opened the pull request:
After removing 20ms in advertiser, all subnetwork beacons are sent as high dense packet of frames with minimal distance between them. That might cause collisions if beacon interval on devices will coincide.

This commit adds an artificial delay between each subnet advertsing beacons.

alxelax reviewed (changes requested):
Did you try to revert commit with test changes according to this topic? https://github.com/zephyrproject-rtos/zephyr/pull/66538#issuecomment-1857548852

Andrewpini commented:
> Did you try to revert commit with test changes according to this topic? [#66538 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/66538#issuecomment-1857548852)

Yes, I tried removing the 500ms delay from the multiple subnet test. It did not pass. Since the test iterates it messes up timing alignment eventually. Maybe it could be refactored to sync better between each iteration?

alxelax commented on subsys/bluetooth/mesh/beacon.c:
```diff
@@ -449,6 +454,26 @@ static bool net_beacon_is_running(void)
+		sub_curr = sub_next;
+	} while (!net_beacon_for_subnet_send(sub_curr));
+
+		k_work_reschedule(&beacon_send_work, K_MSEC(20));
```
alignment issue

alxelax reviewed (commented):
LGTM just need to fix alignment

PavelVPV commented on subsys/bluetooth/mesh/beacon.c:
```diff
@@ -744,6 +768,7 @@ BT_MESH_SUBNET_CB_DEFINE(beacon) = {
 void bt_mesh_beacon_init(void)
 {
 	k_work_init_delayable(&beacon_timer, beacon_send);
+	k_work_init_delayable(&beacon_send_work, beacon_send_work_cb);
```
Can't this be handled in `beacon_send` with some 1 byte variable describing current action?

PavelVPV commented on subsys/bluetooth/mesh/beacon.c:
```diff
@@ -449,6 +454,26 @@ static bool net_beacon_is_running(void)
+		sub_curr = sub_next;
+	} while (!net_beacon_for_subnet_send(sub_curr));
+
+	k_work_reschedule(&beacon_send_work, K_MSEC(20));
```
Does this split beacons it time if advertiser is busy sending other local messages? If local message has network transmit state > 1, it will "block" the main adv instance for more than 20ms. If there are few such messages in the adv pool, this work will add beacons faster than advertiser is able to consume advs. So once the advertiser picks beacons, it will send them one by one without delay.

PavelVPV replied:
> Maybe use the end cb for the sent beacon?

Yes, that is what I though about as well. That should work.

Andrewpini committed:
Bluetooth: Mesh: Add artificial beacon delay

After removing 20ms in advertiser, all subnetwork beacons are sent as high
dense packet of frames with minimal distance between them. That might
cause collisions if beacon interval on devices will coincide.

This commit adds an artificial delay between each subnet advertsing
beacons.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

Andrewpini commented:
Something was wrong with my reset functionality. Need to double check this before merging. I belive it should be reset if the beacons are interrupted mid process, but it looks fine by CI.

alxelax commented on subsys/bluetooth/mesh/beacon.c:
```diff
@@ -449,20 +470,48 @@ static bool net_beacon_is_running(void)
 			return;
 		}
 
 		update_beacon_observation();
```
```
if (!beacon_in_prog) {
    update_beacon_observation();
}
```
you do not need this `goto` construction and initial `if`.

alxelax commented on subsys/bluetooth/mesh/beacon.c:
```diff
@@ -785,11 +841,13 @@ void bt_mesh_beacon_enable(void)
 		bt_mesh_subnet_foreach(subnet_beacon_enable);
 	}
 
+	beacon_in_prog = false;
```
If beacon sending is in progress, it is not required to be reset.
I guess the current implementation is not fully correct. It resets all accumulated data of ongoing process and is able to spam by beacons.

Andrewpini replied:
My thought for why we need to reset is that if the beacon_send progression is interrupted, either by error or other interaction, the `sub_curr` will be in a unknown state. On the next iteration of beacon send, it will then skip whatever subnets that are "below" the `sub_curr`. If you have subnets 0, 1 and 2, a beacon send iteration is interupted while `sub_curr` is pointing on subnet 1, then in the next iteration beacon_send will only advertise subnets 1 and 2.

Andrewpini commented on subsys/bluetooth/mesh/beacon.c:
```diff
@@ -267,14 +283,14 @@ static bool net_beacon_send(struct bt_mesh_subnet *sub, struct bt_mesh_beacon *b
 		(time_diff < BEACON_THRESHOLD(beacon) ||
 		 time_since_last_recv < (10 * MSEC_PER_SEC))) {
-		return false;
+		return ENOMSG;
```
This is not meant as an actual error, just an indication that for this subnet no message will be issued. It is to ensure that `beacons_send_next` continues to the next subnet if no message is issued. I my have misunderstood, but a long time ago I learned that errors are indicated by the `-` in front, and other status codes that is not an error are positive?

alxelax replied:
Fully agree with Pavel. `bool` was suitable here.

Andrewpini commented on subsys/bluetooth/mesh/beacon.c:
```diff
@@ -284,12 +300,17 @@ static bool net_beacon_send(struct bt_mesh_subnet *sub, struct bt_mesh_beacon *b
+static int net_beacon_for_subnet_send(struct bt_mesh_subnet *sub)
 {
-	bool res = true;
+	if (!sub) {
```
At some point it was I think, but maybe not anymore. I can try to remove it.

Andrewpini commented on subsys/bluetooth/mesh/beacon.c:
```diff
@@ -284,12 +300,17 @@ static bool net_beacon_send(struct bt_mesh_subnet *sub, struct bt_mesh_beacon *b
+		return -ENOENT;
+	}
+
+	int err = -ENOMSG;
```
Yes, I think you are right.

PavelVPV commented on subsys/bluetooth/mesh/beacon.c:
```diff
@@ -449,6 +465,31 @@ static bool net_beacon_is_running(void)
+		if (sub_next == sub_first && beacon_send_sub_curr != NULL) {
+			beacon_send_sub_curr = NULL;
+			return false;
+		}
+		beacon_send_sub_curr = sub_next;
+		err = net_beacon_for_subnet_send(beacon_send_sub_curr);
+
+		if (err < 0) {
```
Empty line in a wrong place...
```suggestion
		}

		beacon_send_sub_curr = sub_next;
		err = net_beacon_for_subnet_send(beacon_send_sub_curr);
		if (err < 0) {
```

PavelVPV commented on subsys/bluetooth/mesh/beacon.c:
```diff
@@ -284,12 +300,12 @@ static bool net_beacon_send(struct bt_mesh_subnet *sub, struct bt_mesh_beacon *b
+static int net_beacon_for_subnet_send(struct bt_mesh_subnet *sub)
 {
-	bool res = true;
+	int err = ENOMSG;
```
?

Andrewpini replied:
Sry, forgot it.

Andrewpini commented on subsys/bluetooth/mesh/beacon.c:
```diff
@@ -267,14 +283,14 @@ static bool net_beacon_send(struct bt_mesh_subnet *sub, struct bt_mesh_beacon *b
 		(time_diff < BEACON_THRESHOLD(beacon) ||
 		 time_since_last_recv < (10 * MSEC_PER_SEC))) {
-		return false;
+		return 1;
```
It is evaluated in `beacons_send_next`:

```
		if (err < 0) {
			LOG_ERR("Failed to advertise beacon: err %d", err);
			beacon_send_sub_curr = NULL;
			return false;
		}
	} while (err);
```
if err=1 then it will jump to the next subnet

alxelax replied:
It jumps to the next if it is `not zero`. `1` or `-21` doesn't matter :)

Andrewpini replied:
Ok, so do you want me to return an error code then for this case? It can not be zero.
